File: backend/main-service/src/main/java/com/pethaven/main/model/nosql/Notification.java
package com.pethaven.main.model.nosql;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.util.Map;

@Document(collection = "notifications")
public class Notification {

    @Id
    private String id;
    private String recipientId;
    private String recipientRole;
    private String type; // adoption, vaccination, payment, appointment, shelter, complaint, complaint_status, system
    private String title;
    private String message;
    private boolean isRead;
    private Map<String, String> data;
    private String createdAt;

    public Notification() {}

    // Getters and Setters
    public String getId() { return id; }
    public void setId(String id) { this.id = id; }

    public String getRecipientId() { return recipientId; }
    public void setRecipientId(String recipientId) { this.recipientId = recipientId; }

    public String getRecipientRole() { return recipientRole; }
    public void setRecipientRole(String recipientRole) { this.recipientRole = recipientRole; }

    public String getType() { return type; }
    public void setType(String type) { this.type = type; }

    public String getTitle() { return title; }
    public void setTitle(String title) { this.title = title; }

    public String getMessage() { return message; }
    public void setMessage(String message) { this.message = message; }

    public boolean isRead() { return isRead; }
    public void setRead(boolean read) { isRead = read; }

    public Map<String, String> getData() { return data; }
    public void setData(Map<String, String> data) { this.data = data; }

    public String getCreatedAt() { return createdAt; }
    public void setCreatedAt(String createdAt) { this.createdAt = createdAt; }
}

File: backend/models/Notification.js
import mongoose from 'mongoose';

const notificationSchema = new mongoose.Schema({
  recipientId: { 
    type: mongoose.Schema.Types.ObjectId, 
    ref: 'User',
    required: true 
  },
  recipientRole: { 
    type: String, 
    enum: ['adopter', 'shelter', 'vet', 'veterinarian', 'admin'],
    required: true 
  },
  type: { 
    type: String, 
    enum: ['adoption', 'vaccination', 'payment', 'appointment', 'shelter', 'complaint', 'complaint_status', 'system'],
    required: true 
  },
  title: { type: String, required: true },
  message: { type: String, required: true },
  isRead: { type: Boolean, default: false },
  data: { type: mongoose.Schema.Types.Mixed }, // Additional data (e.g., complaintId, petId, etc.)
  createdAt: { type: Date, default: Date.now },
  readAt: { type: Date }
});

// Index for faster queries
notificationSchema.index({ recipientId: 1, createdAt: -1 });
notificationSchema.index({ recipientRole: 1, isRead: 1 });

export default mongoose.model('Notification', notificationSchema);
